hass_nabucasa/ice_servers.py

```diff
@@ -0,0 +1,137 @@
+"""Manage ICE servers."""
+
+from __future__ import annotations
+
+import asyncio
+from collections.abc import Awaitable, Callable
+import logging
+import time
+from typing import TYPE_CHECKING
+
+from aiohttp import ClientResponseError
+from aiohttp.hdrs import AUTHORIZATION, USER_AGENT
+from webrtc_models import RTCIceServer
+
+if TYPE_CHECKING:
+    from . import Cloud, _ClientT
+
+
+_LOGGER = logging.getLogger(__name__)
+
+
+class IceServers:
+    """Class to manage ICE servers."""
+
+    def __init__(self, cloud: Cloud[_ClientT]) -> None:
+        """Initialize ICE Servers."""
+        self.cloud = cloud
+        self._refresh_task: asyncio.Task | None = None
+        self._ice_servers: list[RTCIceServer] = []
+        self._ice_servers_listener: Callable[[], Awaitable[None]] | None = None
+        self._ice_servers_listener_unregister: list[Callable[[], None]] = []
+
+    async def _async_fetch_ice_servers(self) -> None:
+        """Fetch ICE servers."""
+        if TYPE_CHECKING:
+            assert self.cloud.id_token is not None
+
+        async with self.cloud.websession.get(
+            f"https://{self.cloud.servicehandlers_server}/webrtc/ice_servers",
+            headers={
+                AUTHORIZATION: self.cloud.id_token,
+                USER_AGENT: self.cloud.client.client_name,
+            },
+        ) as resp:
+            resp.raise_for_status()
+
+            self._ice_servers = [
+                RTCIceServer(
+                    urls=item["urls"],
+                    username=item["username"],
+                    credential=item["credential"],
+                )
+                for item in await resp.json()
+            ]
+
+        if self._ice_servers_listener is not None:
+            await self._ice_servers_listener()
+
+    def _get_refresh_sleep_time(self) -> int:
+        """Get the sleep time for refreshing ICE servers."""
+        timestamps = [
+            int(server.username.split(":")[0])
+            for server in self._ice_servers
+            if server.urls.startswith("turn:")
+        ]
+
+        if not timestamps:
+            return 3600  # 1 hour
+
+        # 1 hour before the earliest expiration
+        return min(timestamps) - int(time.time()) - 3600
```

Comment:
Do we need further validation or error handling? Eg could this result in a negative value? Does it matter? Do we need a default value if the calculated value goes beyond some interval?

Reply:
Good point. I added a check to refresh within a timeframe if the expiration time is less than 1 hour.
I don't think we should have any max refresh time check, as the cloud should be responsible for that.